### PR TITLE
[tibber] Add time series support for Tibber prices

### PR DESCRIPTION
**Pull request by soenkekueper**
This PR addeds time series support for future tibber prices to openhab.
The values, that are currenty served as json data, are parsed and served as time series for the current total costs channel.

### CODE REVIEW

File: bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java [hunk 1]
```diff
@@ -267,6 +273,49 @@ public void getURLInput(String url) throws IOException {
...
+        for (JsonElement entry : today) {
+            DateTimeType startsAt = parseDateTime(entry.getAsJsonObject().get("startsAt"));
+            timeSeries.add( //
+                    startsAt.getInstant(), //
+                    new DecimalType(entry.getAsJsonObject().get("total").getAsString()) //
+            );
+            timeSeries.add( //
+                    startsAt.getInstant().plus(1, ChronoUnit.HOURS).minus(1, ChronoUnit.MICROS), //
+                    new DecimalType(entry.getAsJsonObject().get("total").getAsString()) //
+            );
```
Comment (jlaur) [inline bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java:296]: This seems a little odd, why are you adding the same entry again with a new timestamp?

File: bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java [hunk 2]
```diff
@@ -267,6 +273,49 @@ public void getURLInput(String url) throws IOException {
...
+     * 
+     * @param today The prices for today
+     * @param tomorrow The prices for tomorrow.
+     * @return The {@link TimeSeries} with future values.
+     */
+    private TimeSeries buildTimeSeries(JsonArray today, JsonArray tomorrow) {
+        final TimeSeries timeSeries = new TimeSeries(TimeSeries.Policy.REPLACE);
+
+        for (JsonElement entry : today) {
+            DateTimeType startsAt = parseDateTime(entry.getAsJsonObject().get("startsAt"));
```
Comment (jlaur) [inline bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java:288]: It seems unneeded to create a `DateTimeType` just to have a formatted `ZonedDateTime` string parsed. Could you simply use `ZonedDateTime.parse` instead? This is what `DateTimeType` does, except it tries a little harder with different formats. 😉

File: bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java [hunk 3]
```diff
@@ -267,6 +273,49 @@ public void getURLInput(String url) throws IOException {
...
+            timeSeries.add( //
+                    startsAt.getInstant(), //
+                    new DecimalType(entry.getAsJsonObject().get("total").getAsString()) //
+            );
+            timeSeries.add( //
+                    startsAt.getInstant().plus(1, ChronoUnit.HOURS).minus(1, ChronoUnit.MICROS), //
+                    new DecimalType(entry.getAsJsonObject().get("total").getAsString()) //
+            );
+        }
```
Comment (jlaur) [inline bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java:309]: Instead of duplicating this code, could you create the `TimeSeries` outside of this method, and call the method for today and tomorrow respectively, providing the created `TimeSeries`?

File: bundles/org.openhab.binding.tibber/README.md [hunk 4]
```diff
@@ -102,6 +102,7 @@ Tibber API will be auto discovered if provided input is correct.
 
 ## Tomorrow and Today Prices
 
+The today and tomorrow prices are served as time series on the _Current Total_ channel and as json data on the channels _Today prices_ and _Tomorrow prices_.
```
Comment (jlaur) [inline bundles/org.openhab.binding.tibber/README.md:105]: Perhaps it would be better to mention the concrete/actual names of the channels?
```suggestion
The today and tomorrow prices are served as time series on the `current_total` channel and as JSON data on the channels `today_prices` and `tomorrow_prices`.
```

**Review by jlaur — CHANGES_REQUESTED** [review]
Thanks for adding time series support! I have added a few quick comments.

File: bundles/org.openhab.binding.tibber/README.md [hunk 5]
```diff
@@ -77,7 +81,7 @@ Note: Tibber HomeId is retrieved from [developer.tibber.com](https://developer.t
 - If Tibber Pulse is connected, the Tibber API Explorer will report "true" for "realTimeConsumptionEnabled"
 - Copy HomeId from Tibber API Explorer, without quotation marks, and use this in the bindings configuration.
 
-```json
+```
```
Comment (jlaur) [inline bundles/org.openhab.binding.tibber/README.md:84]: Please revert this change for proper syntax highlighting.

File: bundles/org.openhab.binding.tibber/README.md [hunk 6]
```diff
@@ -209,13 +214,13 @@ Example of tomorrow and today prices data structure - an array of tuples:
 
 ### demo.things
 
-```java
+```
```
Comment (jlaur) [inline bundles/org.openhab.binding.tibber/README.md:217]: Please revert this change for proper syntax highlighting.

File: bundles/org.openhab.binding.tibber/README.md [hunk 7]
```diff
@@ -209,13 +214,13 @@ Example of tomorrow and today prices data structure - an array of tuples:
...
 Thing tibber:tibberapi:7cfae492 [ homeid="xxx", token="xxxxxxx" ]
 ```
 
 ### demo.items:
 
-```java
+```
```
Comment (jlaur) [inline bundles/org.openhab.binding.tibber/README.md:223]: Please revert this change for proper syntax highlighting.

File: bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java (hunk 1 above)
Comment (soenkekueper) [reply bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java]: Because this is the semantics of the api: The given price starts at the given time and is valid for exactly one hour (when the next price starts). If the second entry is not present, openhab assumes (and displays) a linear curve, which is wrong. This way it's display just the way i want and how its displayed whitin the tibber app: as stepping function:
Tibber preview
I've added a comment in source code, too.

File: bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java (hunk 2 above)
Comment (soenkekueper) [reply bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java]: fixed.

File: bundles/org.openhab.binding.tibber/README.md (hunk 4 above)
Comment (soenkekueper) [reply bundles/org.openhab.binding.tibber/README.md]: I've just taken the value via C&P from the table above. So remarked, that they're all wrong and corrected that.
Added a column "forecast" so it's clear when mapping the channels and no need to read all the text below :)

File: bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java [hunk 8]
```diff
@@ -267,6 +275,36 @@ public void getURLInput(String url) throws IOException {
...
+            final DecimalType value = new DecimalType(entry.getAsJsonObject().get("total").getAsString());
+            // As the value is valid for exactly one hour and is not linear, add one entry at the beginning
+            // and one at the end of the interval, so it's a step function instead a linear one.
+            timeSeries.add(startsAt, value);
+            timeSeries.add(endsAt, value);
+        }
+    }
+
+    private static Instant parseDateTime(JsonElement element) {
+        return ZonedDateTime.parse(element.getAsString()).toInstant();
```
Comment (jlaur) [inline bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java:305]: This can throw [DateTimeParseException](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/time/format/DateTimeParseException.html), which should be caught so that the binding will not crash on unexpected JSON data.

File: bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java (hunk 1 above)
Comment (jlaur) [reply bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java]: This is how such line charts are generated. You can create a bar chart instead for a better representation. Without time series you would have the same issue when the current price is updated once per hour and looking at a chart back in time. I understand what you are trying to accomplish, but I don't think it's the right way. Logically, this work-around causes these events:
- 14.01.2023 12:00:00.000 1 EUR
- 14.01.2023 12:59:59.999 1 EUR
- 14.01.2023 13:00:00.000 1,1 EUR
- 14.01.2023 13:59:59.999 1,1 EUR

It shouldn't be needed to push an item state change to the event bus when the value didn't change. So I think this problem is in how you created the chart, and it would be the same issue for all line charts. Maybe there are some options for modifing line charts in this regard, but I'm not strong in this area.

File: bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java [hunk 9]
```diff
@@ -267,6 +275,36 @@ public void getURLInput(String url) throws IOException {
...
+     * @return The {@link TimeSeries} with future values.
+     */
+    private TimeSeries buildTimeSeries(JsonArray today, JsonArray tomorrow) {
+        final TimeSeries timeSeries = new TimeSeries(TimeSeries.Policy.REPLACE);
+        mapTimeSeriesEntries(today, timeSeries);
+        mapTimeSeriesEntries(tomorrow, timeSeries);
+        return timeSeries;
+    }
+
+    private static void mapTimeSeriesEntries(JsonArray today, TimeSeries timeSeries) {
```
Comment (jlaur) [inline bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java:292]:
```suggestion
    private static void mapTimeSeriesEntries(JsonArray prices, TimeSeries timeSeries) {
```

(or similar, since "today" is now misleading)

**Review by jlaur — CHANGES_REQUESTED** [review]
Thanks for the updates. I have provided some additional feedback.

File: bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java (hunk 1 above)
Comment (jlaur) [reply bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java]: @J-N-K - do you have an opinion on this matter?

File: bundles/org.openhab.binding.tibber/README.md (hunk 7 above)
Comment (soenkekueper) [reply bundles/org.openhab.binding.tibber/README.md]: done

File: bundles/org.openhab.binding.tibber/README.md (hunk 6 above)
Comment (soenkekueper) [reply bundles/org.openhab.binding.tibber/README.md]: done

File: bundles/org.openhab.binding.tibber/README.md (hunk 5 above)
Comment (soenkekueper) [reply bundles/org.openhab.binding.tibber/README.md]: Done. as this is not a valid json and IJ displayed a lot of errors, i've removed them.

File: bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java (hunk 9 above)
Comment (soenkekueper) [reply bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java]: done.

File: bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java (hunk 1 above)
Comment (soenkekueper) [reply bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java]: I've found an option for the apache eChart to display a step line:
https://echarts.apache.org/en/option.html#series-line.step
So this fixes the issue a bette way.
Thanks for your input!

File: bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java (hunk 8 above)
Comment (soenkekueper) [reply bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java]: done.

File: bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java [hunk 10]
```diff
@@ -267,6 +274,42 @@ public void getURLInput(String url) throws IOException {
...
+            timeSeries.add(startsAt, value);
+        }
+    }
+
+    @Nullable
+    private Instant parseDateTime(JsonElement element) {
+        try {
+            return ZonedDateTime.parse(element.getAsString()).toInstant();
+        } catch (DateTimeParseException e) {
+            logger.warn("Could not parse timestamp from response.", e);
```
Comment (jlaur) [inline bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java:308]: This will log a warning per entry. As it is most likely that all entries will be either valid or invalid, not just some of them, I would suggest to catch this exception in the `getURLInput` method. You could even add the exception in 213 and handle it the same way as `JsonSyntaxException` already caught there. That will exit the loop fast and also consolidate the error handling while also setting the thing status to offline.

Consider also if you need this private method at all, since it's only called from line 293, and it's a one-liner. Perhaps ZonedDateTime.parse could be called directly from `mapTimeSeriesEntries`.

**Review by jlaur — CHANGES_REQUESTED** [review]
Otherwise LGTM.

**Comment by soenkekueper** [comment]
I've rasied a PR to add the tibber icon in docs, as it is missing:
https://github.com/openhab/openhab-docs/pull/2212

**Review by jlaur — APPROVED** [review]
LGTM. I'll give @kjoglum a chance to have a look as well.

**Comment by kjoglum** [comment]
Looks ok to me.

**Comment by jlaur** [comment]
> Looks ok to me.

Thanks for checking, let's merge then.
